Run scripts with `sudo` when `become` is on

carlosrodfern opened the pull request:
Implements: #2366

carlosrodfern commented:
@happz ,
I added the changes in the schemas and also a test on a container.
Could you please check this PR? Thank you.

carlosrodfern commented:
> Please lets decide about [#2366 (comment)](https://github.com/teemtee/tmt/issues/2366#issuecomment-1754809139) first (and again apologies for being late to the party)
> 
> Other nitpick is that we could use the GuestFact `is_superuser` and do sudo only when necessary.

Hi @lukaszachy , thank you for the feedback!
I believe there was already some MR to do "sudo" when necessary in some steps, but not sure if that would work for prepare/finish. I'll work on the `become` also to get applied to prepare and finish. I think it makes a lot of sense to apply it to prepare and finish as well, since that would be what a user expects from a flag like that.

carlosrodfern commented:
@lukaszachy @happz , I pushed a new commit with the `become` handling in the prepare and finish step and also cleaned up the tests and added new tests for this change.

carlosrodfern commented:
@lukaszachy @happz , I do have a question about this statement in the Shell.py in prepare and finish:

```python
if not self.is_dry_run and self.step.plan.worktree:
```

The script run were already setup to run even when `is_dry_run`, but those scripts may depend on the steps which are condition by that same flag. Is this a bug or am I missing something?

Also, I'm trying to find out in what case `self.step.plan.worktree` is `None`. My code for prepare and finish would fail if `is_dry_run` is truthy or when `self.step.plan.worktree` is `None`.

carlosrodfern commented:
> @lukaszachy @happz , I do have a question about this statement in the Shell.py in prepare and finish:
> 
> ```python
> if not self.is_dry_run and self.step.plan.worktree:
> ```
> 
> The script run were already setup to run even when `is_dry_run`, but those scripts may depend on the statements which are conditioned by that same flag. Is this a bug or am I missing something?
> 
> Also, I'm trying to find out in what case `self.step.plan.worktree` is `None`. My code for prepare and finish would fail if `is_dry_run` is truthy or when `self.step.plan.worktree` is `None`.

Actually, that conditional, is only in the prepare step, ...but still, `self.step.plan.worktree` can be `None` even in the finish. I'm looking into it.

happz commented:
> > > @lukaszachy @happz , I do have a question about this statement in the Shell.py in prepare and finish:
> > > ```python
> > > if not self.is_dry_run and self.step.plan.worktree:
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > The script run were already setup to run even when `is_dry_run`, but those scripts may depend on the statements which are conditioned by that same flag. Is this a bug or am I missing something?
> > > Also, I'm trying to find out in what case `self.step.plan.worktree` is `None`. My code for prepare and finish would fail if `is_dry_run` is truthy or when `self.step.plan.worktree` is `None`.
> > 
> > 
> > Actually, that conditional is only in the prepare step, ...but still, `self.step.plan.worktree` can be `None` even in the finish. I'm looking into it.
> 
> Yeah, it looks to me that `worktree` is indeed necessary to run any finish and prepare script, since there can be a reference to a file, e.g., `script: scripts/prepare.sh`.
> 
> Regarding the `dry_run`, should the entire `go(..)` in `PrepareShell` and `FinishShell` be conditioned with the `is_dry_run` flag? Something like the `PrepareInstall`:
> 
> ```python
> # Nothing to do in dry mode
>         if self.is_dry_run:
>             return
> ```

I'd say "no" here. IIRC, `guest.execute()` does some magic under the hood which, in dry-run mode, does not execute any commands, just logs them. It's not perfect, and yep, some commands may depend on the outcome of previous ones, but I wouldn't change this behavior in this PR.

carlosrodfern commented:
> > > > @lukaszachy @happz , I do have a question about this statement in the Shell.py in prepare and finish:
> > > > ```python
> > > > if not self.is_dry_run and self.step.plan.worktree:
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > The script run were already setup to run even when `is_dry_run`, but those scripts may depend on the statements which are conditioned by that same flag. Is this a bug or am I missing something?
> > > > Also, I'm trying to find out in what case `self.step.plan.worktree` is `None`. My code for prepare and finish would fail if `is_dry_run` is truthy or when `self.step.plan.worktree` is `None`.
> > > 
> > > 
> > > Actually, that conditional is only in the prepare step, ...but still, `self.step.plan.worktree` can be `None` even in the finish. I'm looking into it.
> > 
> > 
> > Yeah, it looks to me that `worktree` is indeed necessary to run any finish and prepare script, since there can be a reference to a file, e.g., `script: scripts/prepare.sh`.
> > Regarding the `dry_run`, should the entire `go(..)` in `PrepareShell` and `FinishShell` be conditioned with the `is_dry_run` flag? Something like the `PrepareInstall`:
> > ```python
> > # Nothing to do in dry mode
> >         if self.is_dry_run:
> >             return
> > ```
> 
> I'd say "no" here. IIRC, `guest.execute()` does some magic under the hood which, in dry-run mode, does not execute any commands, just logs them. It's not perfect, and yep, some commands may depend on the outcome of previous ones, but I wouldn't change this behavior in this PR.

I just saw your comment. I had pushed a commit with a proposal. I'll remove my commit, then. What about the `worktree` possibly being `None` and depending on it to run shell scripts with references to shell files (e.g. `scripts/prepare.sh`)?

carlosrodfern commented:
> > > > > @lukaszachy @happz , I do have a question about this statement in the Shell.py in prepare and finish:
> > > > > ```python
> > > > > if not self.is_dry_run and self.step.plan.worktree:
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > The script run were already setup to run even when `is_dry_run`, but those scripts may depend on the statements which are conditioned by that same flag. Is this a bug or am I missing something?
> > > > > Also, I'm trying to find out in what case `self.step.plan.worktree` is `None`. My code for prepare and finish would fail if `is_dry_run` is truthy or when `self.step.plan.worktree` is `None`.
> > > > 
> > > > 
> > > > Actually, that conditional is only in the prepare step, ...but still, `self.step.plan.worktree` can be `None` even in the finish. I'm looking into it.
> > > 
> > > 
> > > Yeah, it looks to me that `worktree` is indeed necessary to run any finish and prepare script, since there can be a reference to a file, e.g., `script: scripts/prepare.sh`.
> > > Regarding the `dry_run`, should the entire `go(..)` in `PrepareShell` and `FinishShell` be conditioned with the `is_dry_run` flag? Something like the `PrepareInstall`:
> > > ```python
> > > # Nothing to do in dry mode
> > >         if self.is_dry_run:
> > >             return
> > > ```
> > 
> > 
> > I'd say "no" here. IIRC, `guest.execute()` does some magic under the hood which, in dry-run mode, does not execute any commands, just logs them. It's not perfect, and yep, some commands may depend on the outcome of previous ones, but I wouldn't change this behavior in this PR.
> 
> I just saw your comment. I had pushed a commit with a proposal. I'll remove my commit, then. What about the `worktree` possibly being `None` and depending on it to run shell scripts with references to shell files (e.g. `scripts/prepare.sh`)?

For what I can understand `worktree` is not initialized during a `clean` run only, so I think it might make sense to skip the prepare and finish if `worktree` is `None`.

happz commented:
> > ...
> > 
> > I just saw your comment. I had pushed a commit with a proposal. I'll remove my commit, then. What about the `worktree` possibly being `None` and depending on it to run shell scripts with references to shell files (e.g. `scripts/prepare.sh`)?
> 
> For what I can understand `worktree` is not initialized during a `clean` run only, so I think it might make sense to skip the prepare and finish if `worktree` is `None`.

Hmm, I'm not 100 % sure in this area, handling of dry-run is a bit shady. We should have better detection, more focused on what rather than how, but that'd be for other patches. For now, I for one would not skip whole `go()` when `worktree` is unset, basically because I'd fear the unexpected consequences down the stream. This would mean `guest.execute()` calls would not happen, which means no dry-run logging of those commands in logs, who knows what else could happen.

@psss @lukaszachy what do you think?

carlosrodfern commented:
> > > ...
> > > I just saw your comment. I had pushed a commit with a proposal. I'll remove my commit, then. What about the `worktree` possibly being `None` and depending on it to run shell scripts with references to shell files (e.g. `scripts/prepare.sh`)?
> > 
> > 
> > For what I can understand `worktree` is not initialized during a `clean` run only, so I think it might make sense to skip the prepare and finish if `worktree` is `None`.
> 
> Hmm, I'm not 100 % sure in this area, handling of dry-run is a bit shady. We should have better detection, more focused on what rather than how, but that'd be for other patches. For now, I for one would not skip whole `go()` when `worktree` is unset, basically because I'd fear the unexpected consequences down the stream. This would mean `guest.execute()` calls would not happen, which means no dry-run logging of those commands in logs, who knows what else could happen.
> 
> @psss @lukaszachy what do you think?

I ended up not skipping on dry run and I did test it and all I had to do was to skip the chmod (see my last commit).

However, `worktree` is a dependency for prepare and finish to complete. It was a weak dependency before because it was only needed when the prepare or finish shell had a reference to a script in the work tree, but with sudo, where I need to create a wrapper script, it comes a strong dependency. It appears that `worktree` is only `None` when running `clean`. Now, this is the interesting thing, browsing the code, and running some verification tests, I don't think `clean` ever runs the prepare or finish step, so, it appears that `worktree` is always set when it hits those functions.
So, it may be totally OK to leave it like that, but whoever comes later to read the code will be confused for what it sees.

lukaszachy commented:
> @lukaszachy what do you think?

I wish we could get rid of dry-run :) Too many additional branches with  too little benefit.

carlosrodfern commented on tmt/steps/execute/internal.py:
```diff
@@ -223,7 +223,11 @@ def execute(
         # Prepare the actual remote command
-        remote_command = ShellScript(f'./{test_wrapper_filename}')
+        remote_command: ShellScript
+        if guest.become:
```
Thank you @lukaszachy for the feedback. I’ll prepare a commit with all those changes.

carlosrodfern replied:
Pushed new commit with this fix

carlosrodfern commented on tmt/steps/finish/shell.py:
```diff
@@ -67,8 +69,27 @@ def go(
+                destination=finish_wrapper_path,
+                options=["-s", "-p", "--chmod=755"])
+            command: ShellScript
+            if guest.become:
```
Pushed new commit with this fix

carlosrodfern commented on tmt/steps/prepare/shell.py:
```diff
@@ -86,8 +88,30 @@ def go(
+                destination=prepare_wrapper_path,
+                options=["-s", "-p", "--chmod=755"])
+            command: ShellScript
+            if guest.become:
```
Pushed new commit with this fix

carlosrodfern commented:
I collapsed all the commits into one

carlosrodfern commented:
> > > > ...
> > > > I just saw your comment. I had pushed a commit with a proposal. I'll remove my commit, then. What about the `worktree` possibly being `None` and depending on it to run shell scripts with references to shell files (e.g. `scripts/prepare.sh`)?
> > > 
> > > 
> > > For what I can understand `worktree` is not initialized during a `clean` run only, so I think it might make sense to skip the prepare and finish if `worktree` is `None`.
> > 
> > 
> > Hmm, I'm not 100 % sure in this area, handling of dry-run is a bit shady. We should have better detection, more focused on what rather than how, but that'd be for other patches. For now, I for one would not skip whole `go()` when `worktree` is unset, basically because I'd fear the unexpected consequences down the stream. This would mean `guest.execute()` calls would not happen, which means no dry-run logging of those commands in logs, who knows what else could happen.
> > @psss @lukaszachy what do you think?
> 
> I ended up not skipping on dry run and I did test it and all I had to do was to skip the chmod (see my last commit). So it works fine now without much changes with the dry-run mode.
> 
> However, `worktree` is a dependency for prepare and finish to complete (assuming no dry run mode). It was a weak dependency before because it was only needed when the prepare or finish shell had a reference to a script in the work tree, but with sudo, where I need to create a wrapper script, it becomes a strong dependency. It appears that `worktree` is only `None` when running `clean`. Now, this is the interesting thing, browsing the code, and running some verification tests, I don't think `clean` ever runs the prepare or finish step, so, it appears that `worktree` is always set when it hits those functions. So, it may be totally OK to leave it like that, but whoever comes later to read the code will be confused for what they see.

@happz @lukaszachy , I just pushed a commit with the cleanup I'm proposing from this discussion. The commit explains the reasoning for future reference. I did run some of the tmt tests locally and they were passing. Please let me know what you think of if you want me to follow another route.
Thank you.

happz commented on tmt/steps/prepare/shell.py:
```diff
@@ -81,13 +86,34 @@ def go(
                     filename_base=filename_base))
 
+
+        prepare_wrapper_filename = f'{PREPARE_WRAPPER_FILENAME}.{guest.name}'
```
The name should follow the same pattern as `filename_base` above, i.e. the filename base, specific for the purpose, followed by phase and guest *safe* names: `f'{PREPARE_WRAPPER_FILENAME}-{self.safe_name}-{guest.safe_name}'`.

Please, follow the same pattern, and I'll submit a patch adding a helper so we don't have to worry about such low-level details in the future.

happz replied:
Yep, we need a helper function for this, I just noticed internal execute plugin follows a different pattern :)

happz replied:
FTR, it's https://github.com/teemtee/tmt/pull/2424. Don't bother waiting for it, unless you really want to - I scheduled it for 1.30 milestone.

carlosrodfern replied:
Pushed new commit with the fix. I also fixed the "finish" wrapper script name with the same pattern. Thank you @happz !

carlosrodfern commented:
I found a new issue with the change: https://github.com/teemtee/tmt/commit/a6ad1ec5d7afba6a902f08e22cd5306fa073aa11

This is an output after I added extra information to the `TEST_WRAPPER_TEMPLATE` var.
```
                cmd: echo "pass"
                out: ENTER
                out: total 0
                out: 0 drwxr-xr-t. 1 root root 14 Oct 24 19:27 .
                out: 0 drwxrwxrwt. 1 root root 40 Oct 24 19:27 ..
                out: 0 drwxr-xr-t. 1 root root 48 Oct 24 19:27 run-033
                out: flock: cannot open lock file /var/tmp/tmt/tmt-test.pid.lock: Permission denied
```
So the lock file is created by a script that wraps the remote command, which is run with `sudo`. The targeted environment gets the `/var/tmp/tmt` folder created with `root` permissions, so the `flock` doesn't have the permissions to manage that lock file. I'm looking into it.

happz commented:
> I found a new issue with the change: [a6ad1ec](https://github.com/teemtee/tmt/commit/a6ad1ec5d7afba6a902f08e22cd5306fa073aa11)
> 
> This is an output after I added extra information to the `TEST_WRAPPER_TEMPLATE` var.
> 
> ```
>                 cmd: echo "pass"
>                 out: ENTER
>                 out: total 0
>                 out: 0 drwxr-xr-t. 1 root root 14 Oct 24 19:27 .
>                 out: 0 drwxrwxrwt. 1 root root 40 Oct 24 19:27 ..
>                 out: 0 drwxr-xr-t. 1 root root 48 Oct 24 19:27 run-033
>                 out: flock: cannot open lock file /var/tmp/tmt/tmt-test.pid.lock: Permission denied
> ```
> 
> So the lock file is created by a script that wraps the remote command, which is run with `sudo`. The targeted environment gets the `/var/tmp/tmt` folder created with `root` permissions, so the `flock` doesn't have the permissions to manage that lock file. I'm looking into it.

Hmmm, at first glance, this looks like a bug in reboot handling. I'm afraid we don't have a test case for this particular scenario (`user` set), and we should get one ASAP to verify that `user` is still supported. Or we can learn to live with a regression... I see no other way than adding `sudo` to lock operations when `guest.facts.is_superuser is not True`. I'll look into fixing this tomorrow morning.

carlosrodfern commented:
> > I found a new issue with the change: [a6ad1ec](https://github.com/teemtee/tmt/commit/a6ad1ec5d7afba6a902f08e22cd5306fa073aa11)
> > This is an output after I added extra information to the `TEST_WRAPPER_TEMPLATE` var.
> > ```
> >                 cmd: echo "pass"
> >                 out: ENTER
> >                 out: total 0
> >                 out: 0 drwxr-xr-t. 1 root root 14 Oct 24 19:27 .
> >                 out: 0 drwxrwxrwt. 1 root root 40 Oct 24 19:27 ..
> >                 out: 0 drwxr-xr-t. 1 root root 48 Oct 24 19:27 run-033
> >                 out: flock: cannot open lock file /var/tmp/tmt/tmt-test.pid.lock: Permission denied
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So the lock file is created by a script that wraps the remote command, which is run with `sudo`. The targeted environment gets the `/var/tmp/tmt` folder created with `root` permissions, so the `flock` doesn't have the permissions to manage that lock file. I'm looking into it.
> 
> Hmmm, at first glance, this looks like a bug in reboot handling. I'm afraid we don't have a test case for this particular scenario (`user` set), and we should get one ASAP to verify that `user` is still supported. Or we can learn to live with a regression... I see no other way than adding `sudo` to lock operations when `guest.facts.is_superuser is not True`. I'll look into fixing this tomorrow morning.

As I read in the code the `/var/tmp/tmt` is expected to be multiuser (777), which is the case in the host machine, however, in the container, it is `drwxr-xr-t`. If that permission discrepancy is fixed, even if the guest `/var/tmp/tmt` is owned by `root`, the lock file part should work.

happz commented:
I will have a couple of other fixes as well, on top of the permissions, but, what if `sudo` is not needed and the permissions are broken because of the `mkdir` in `TEST_WRAPPER_TEMPLATE`? Could you give a chance to the change below, without `sudo` involved?

```diff
-mkdir -p "$(dirname $TMT_TEST_PIDFILE_LOCK)"
+mkdir -p -m 01777 "$(dirname $TMT_TEST_PIDFILE_LOCK)"
```

I think the issue might be actually caused by this command: if the `/var/tmp/tmt` directory does not exist, it's created to hold the pidfile lock, and the permissions are not what tmt uses when creates the directory on its own. The directory might not exist, it can easily happen if tmt is told to use some other directory as a workdir (`--id /var/tmp/tmt.foobar123`) - in that case, tmt does not bother creating `/var/tmp/tmt`, and `mkdir` is the first one creating it.

Maybe we won't need `sudo` after all - it's a bit too late for me, I'm struggling to catch up, but it feels like creating `/var/tmp/tmt` with open-enough permissions should unblock this.

carlosrodfern commented:
> I will have a couple of other fixes as well, on top of the permissions, but, what if `sudo` is not needed and the permissions are broken because of the `mkdir` in `TEST_WRAPPER_TEMPLATE`? Could you give a chance to the change below, without `sudo` involved?
> 
> ```diff
> -mkdir -p "$(dirname $TMT_TEST_PIDFILE_LOCK)"
> +mkdir -p -m 1777 "$(dirname $TMT_TEST_PIDFILE_LOCK)"
> ```
> 
> I think the issue might be actually caused by this command: if the `/var/tmp/tmt` directory does not exist, it's created to hold the pidfile lock, and the permissions are not what tmt uses when creates the directory on its own. The directory might not exist, it can easily happen if tmt is told to use some other directory as a workdir (`--id /var/tmp/tmt.foobar123`) - in that case, tmt does not bother creating `/var/tmp/tmt`, and `mkdir` is the first one creating it.
> 
> Maybe we won't need `sudo` after all - it's a bit too late for me, I'm struggling to catch up, but it feels like creating `/var/tmp/tmt` with open-enough permissions should unblock this.

Adding `-m 1777` didn't work. The code appears to work as-is, without the `sudo flock` on libvirst guests (at least the `become` tests). The problem is specific to containers. See how the folder is being placed in the container (this is running podman as my local user account):

```
Run command: podman run --security-opt seccomp=unconfined --name tmt-038-McwLWkTK -v /var/tmp/tmt/run-038/test_inline_as_root:/var/tmp/tmt/run-038/test_inline_as_root:z -itd --user user localhost/become-container-test:latest
```

This quick change just to validate this issue makes the test with container provision work:
```
diff --git a/tmt/steps/execute/internal.py b/tmt/steps/execute/internal.py
index c00a9c01..3f2d92f0 100644
--- a/tmt/steps/execute/internal.py
+++ b/tmt/steps/execute/internal.py
@@ -205,10 +205,8 @@ class ExecuteInternal(tmt.steps.execute.ExecutePlugin[ExecuteInternalData]):
         assert self.parent is not None
         assert isinstance(self.parent, tmt.steps.execute.Execute)
 
-        environment['TMT_TEST_PIDFILE'] = str(
-            effective_pidfile_root() / TEST_PIDFILE_FILENAME)
-        environment['TMT_TEST_PIDFILE_LOCK'] = str(
-            effective_pidfile_root() / TEST_PIDFILE_LOCK_FILENAME)
+        environment['TMT_TEST_PIDFILE'] = "/tmp/test.pid"
+        environment['TMT_TEST_PIDFILE_LOCK'] = "/tmp/test.pid.lock"
```

So it looks like the solution may be to change the location of the lock when on containers, or to change the way the folders in containers are handled. I'll dig into it.

happz commented:
Ok, would changing (the default) value of the pidfile directory to `/var/tmp`, or even `/tmp`? It's not necessary for the pidfile directory to be shared with other tmt content. It might be enough to unblock you while not harming any other use case, and `/var/tmp/tmt` permissions can be fixed as a standalone issue.

carlosrodfern commented:
> Ok, would changing (the default) value of the pidfile directory to `/var/tmp`, or even `/tmp`? It's not necessary for the pidfile directory to be shared with other tmt content. It might be enough to unblock you while not harming any other use case, and `/var/tmp/tmt` permissions can be fixed as a standalone issue.
> 
> ```diff
> diff --git a/tmt/steps/execute/internal.py b/tmt/steps/execute/internal.py
> index 09280ced..24fda7a3 100644
> --- a/tmt/steps/execute/internal.py
> +++ b/tmt/steps/execute/internal.py
> @@ -24,6 +24,7 @@ from tmt.utils import EnvironmentType, Path, ShellScript, Stopwatch, field
>  TEST_PIDFILE_FILENAME = 'tmt-test.pid'
>  TEST_PIDFILE_LOCK_FILENAME = f'{TEST_PIDFILE_FILENAME}.lock'
>  
> +TEST_PIDFILE_DEFAULT_DIRECTORY = Path('/var/tmp')
>  
>  def effective_pidfile_root() -> Path:
>      """
> @@ -36,7 +37,7 @@ def effective_pidfile_root() -> Path:
>      if 'TMT_TEST_PIDFILE_ROOT' in os.environ:
>          return Path(os.environ['TMT_TEST_PIDFILE_ROOT'])
>  
> -    return tmt.utils.effective_workdir_root()
> +    return TEST_PIDFILE_DEFAULT_DIRECTORY
>  
>  
>  TEST_WRAPPER_FILENAME = 'tmt-test-wrapper.sh'
> ```

I just did a test changing it to `/var/tmp` it it worked.

> Ok, would changing (the default) value of the pidfile directory to `/var/tmp`, or even `/tmp`? It's not necessary for the pidfile directory to be shared with other tmt content. It might be enough to unblock you while not harming any other use case, and `/var/tmp/tmt` permissions can be fixed as a standalone issue.
> 
> ```diff
> diff --git a/tmt/steps/execute/internal.py b/tmt/steps/execute/internal.py
> index 09280ced..24fda7a3 100644
> --- a/tmt/steps/execute/internal.py
> +++ b/tmt/steps/execute/internal.py
> @@ -24,6 +24,7 @@ from tmt.utils import EnvironmentType, Path, ShellScript, Stopwatch, field
>  TEST_PIDFILE_FILENAME = 'tmt-test.pid'
>  TEST_PIDFILE_LOCK_FILENAME = f'{TEST_PIDFILE_FILENAME}.lock'
>  
> +TEST_PIDFILE_DEFAULT_DIRECTORY = Path('/var/tmp')
>  
>  def effective_pidfile_root() -> Path:
>      """
> @@ -36,7 +37,7 @@ def effective_pidfile_root() -> Path:
>      if 'TMT_TEST_PIDFILE_ROOT' in os.environ:
>          return Path(os.environ['TMT_TEST_PIDFILE_ROOT'])
>  
> -    return tmt.utils.effective_workdir_root()
> +    return TEST_PIDFILE_DEFAULT_DIRECTORY
>  
>  
>  TEST_WRAPPER_FILENAME = 'tmt-test-wrapper.sh'
> ```

That solution works. Do you want me to include it in this PR or do you prefer to address it in the other PR you are working on with other fixes? I'm good either way.

happz commented:
> > Ok, would changing (the default) value of the pidfile directory to `/var/tmp`, or even `/tmp`? It's not necessary for the pidfile directory to be shared with other tmt content. It might be enough to unblock you while not harming any other use case, and `/var/tmp/tmt` permissions can be fixed as a standalone issue.
> > ```diff
> > diff --git a/tmt/steps/execute/internal.py b/tmt/steps/execute/internal.py
> > index 09280ced..24fda7a3 100644
> > --- a/tmt/steps/execute/internal.py
> > +++ b/tmt/steps/execute/internal.py
> > @@ -24,6 +24,7 @@ from tmt.utils import EnvironmentType, Path, ShellScript, Stopwatch, field
> >  TEST_PIDFILE_FILENAME = 'tmt-test.pid'
> >  TEST_PIDFILE_LOCK_FILENAME = f'{TEST_PIDFILE_FILENAME}.lock'
> >  
> > +TEST_PIDFILE_DEFAULT_DIRECTORY = Path('/var/tmp')
> >  
> >  def effective_pidfile_root() -> Path:
> >      """
> > @@ -36,7 +37,7 @@ def effective_pidfile_root() -> Path:
> >      if 'TMT_TEST_PIDFILE_ROOT' in os.environ:
> >          return Path(os.environ['TMT_TEST_PIDFILE_ROOT'])
> >  
> > -    return tmt.utils.effective_workdir_root()
> > +    return TEST_PIDFILE_DEFAULT_DIRECTORY
> >  
> >  
> >  TEST_WRAPPER_FILENAME = 'tmt-test-wrapper.sh'
> > ```
> 
> That solution works. Do you want me to include it in this PR or do you prefer to address it in the other PR you are working on with other fixes? I'm good either way.

Another PR it will be. I'll make one in the evening.

happz commented:
The PR changing test pidfile location, https://github.com/teemtee/tmt/pull/2432

lukaszachy reviewed (commented):
We seem to miss update in 'minute' schema -- help shows 'become' is an available option there as well.

I ran into problem when I ran 'as root' first followed by '-u user' -> the `/var/tmp/tmt` is created without o+rwX so rsync failed horribly but it is not the problem of this PR.

carlosrodfern commented:
> We seem to miss update in 'minute' schema -- help shows 'become' is an available option there as well.
> 
> I ran into problem when I ran 'as root' first followed by '-u user' -> the `/var/tmp/tmt` is created without o+rwX so rsync failed horribly but it is not the problem of this PR.

I'll look into the `minute` schema, and also I'm assuming it is the `mrack` provision, so I'll look into that code as well to see what may be lacking to support `become`.

carlosrodfern commented:
@happz ,
I updated my PR with your fix and it passes the tests now. Thank you!

@lukaszachy, I added the `become` to the `minute` schema. It looks like that's all that is needed to be done beside the issue with the permissions.

lukaszachy reviewed (approved):
Thank you very much.

lukaszachy commented:
/packit test

psss reviewed (approved):
Thanks for implementing this! Looks good. Proposing just a couple of minor changes in 31df8bf1. @carlosrodfern, does it look ok to you?

psss commented:
/packit test

psss commented:
/packit test
